crates/bevy_ecs/src/query/iter.rs

```diff
@@ -329,36 +328,29 @@ where
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        if K == 0 {
-            return (0, Some(0));
-        }
-
-        let max_size: usize = self
-            .query_state
-            .matched_archetype_ids
-            .iter()
-            .map(|id| self.archetypes[*id].len())
-            .sum();
-
-        if max_size < K {
-            return (0, Some(0));
-        }
-        if max_size == K {
-            return (1, Some(1));
-        }
-
         // binomial coefficient: (n ; k) = n! / k!(n-k)! = (n*n-1*...*n-k+1) / k!
         // See https://en.wikipedia.org/wiki/Binomial_coefficient
         // See https://blog.plover.com/math/choose.html for implementation
         // It was chosen to reduce overflow potential.
         fn choose(n: usize, k: usize) -> Option<usize> {
+            if k > n || n == 0 {
+                return Some(0);
+            }
+            let k = k.min(n - k);
             let ks = 1..=k;
-            let ns = (n - k + 1..=n).rev();
+            let ns = (n + 1 - k..=n).rev();
             ks.zip(ns)
                 .try_fold(1_usize, |acc, (k, n)| Some(acc.checked_mul(n)? / k))
         }
-        let smallest = K.min(max_size - K);
-        let max_combinations = choose(max_size, smallest);
+        // sum_i=0..k choose(cursors[i].remaining, k-i)
+        let max_combinations = self
+            .cursors
+            .iter()
+            .enumerate()
+            .try_fold(0, |acc, (i, cursor)| {
+                let n = cursor.remaining(self.tables, self.archetypes);
+                Some(acc + choose(n, K - i)?)
+            });
```

Comment:
Note this multiplies by K the runtime of `size_hint` in worst case scenarios, which is significant, since we are computing the binomial. But a few things to know:
* When the iterator has just been initialized (ie: most common use case) this is little different than before this PR, since only the first element of `cursors` has a value of N, and all other of 0 (which has 0 runtime in `choose`
* The worst case is when the iterator has a single element removed, since the cursor remaining count is `[N, N-1, …, N-K]`.

I also I'm just assuming that during monomorphization, this codes gets optimized out when this is not an `ExactSizeIterator` due to the `if archetype_query` (where `archetype_query` can be deduced at compile time).